text/0000-group-additions.md

```diff
@@ -0,0 +1,189 @@
+- Author(s): [jkutner](@jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This proposal describes a new feature in `project.toml` and `pack` that would allow buildpacks to be injected into the middle, beginning, or end of the groups defined in an `order.toml` provided by a builder.
+
+# Definitions
+[definitions]: #definitions
+
+* **injected buildpack** - a buildpack that is added to the beginning, middle, or end of a group.
+* **requisite buildpack** - a buildpack that an injected buildpack depends on to insert in the middle of a group.
+
+# Motivation
+[motivation]: #motivation
+
+Many buildpack users what the default groups provided by a builder image, but need to inject a single buildpack into the middle, begining, or end of those groups. Today, there is no officially supported way to do this without explicitly specifying all of the builpacks in the group or using an undocumented feature like `from=builder` (but `from=builder` does not allow for injecting a buildpack in the middle of a group). Most recently, we've see this recommended by the [Instana Buildpacks](https://github.com/instana/instana-buildpacks/tree/main/google-cloud-platform/cloud-run).
+
+In many of cases where a buildpack needs to be injected into the middle of a group, the injected buildpack either depends on some other buildpack, or must preceed some other buildpack. For this reason, we will divide the proposed solution into two features: one for injecting in the middle of a group, and one for injecting at the beginning or end.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Injecting in the middle of a group
+
+Buildpacks users can specify an injected buildpack after another buildpack in a group by adding something like the following to their `project.toml`:
+
+```toml
+[[build.buildpacks]]
+id = "<buildpack ID>"
+before = "<buildpack ID>"
+```
+
+Or they can inject a buildpack after another another buildpack in a group by adding something like the following to their `project.toml`:
+
+```toml
+[[build.buildpacks]]
+id = "<buildpack ID>"
+after = "<buildpack ID>"
+```
+
+The `after` and `before` keys define a requisite buildpack, which the injected buildpack will run either after or before accordingly.
+
+If injected buildpacks need to conditionally run after more than one buildpack, the user can put the `after` and `before` keys in a `[[build.buildpacks.or]]` array of tables.
+
+## Injecting at the beginning or end of a group
+
+A user can also append buildpacks to the beginning of a group with the following configuration in `project.toml`:
+
+```toml
+[[build.pre.buildpacks]]
+id = "<buildpack ID>"
+```
+
+Or append buildpacks to the end of a group:
+
+```toml
+[[build.post.buildpacks]]
+id = "<buildpack ID>"
+```
+
+With the `pack` CLI, the `pre` and `post` buildpacks can be added with the `--pre-buildpack` and `--post-buildack` options.
```

Comment:
Do we already use the terminology `pre` or `post` anywhere? What do you think of `start`/`end` or `first`/`last`?